File: 백준/Silver/1926. 그림/그림.py

import sys
input = sys.stdin.readline
from collections import deque

def bfs(i,j):
    deq = deque()
    deq.append((i,j))
    dx = [0,0,-1,1]
    dy = [-1,1,0,0]
    graph[i][j] = 0
    cnt = 1

    while deq:
        x,y = deq.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx < 0 or nx > (n-1) or ny < 0 or ny > (m-1):
                continue
            if graph[nx][ny] == 1:
                cnt += 1
                graph[nx][ny] = 0
                deq.append((nx,ny))
    
    return cnt

n,m = map(int,input().split())
graph = [list(map(int,input().split())) for _ in range(n)]
cnt = 0
max_picture = 0

for i in range(n):
    for j in range(m):
        if graph[i][j] == 1:
            cnt += 1
            max_picture = max(max_picture, bfs(i,j))

print(cnt)
print(max_picture)

File: 백준/Silver/1926. 그림/그림.java
import java.util.*;
import java.io.*;

public class Main {
    
    static int n,m;
    static int cnt = 0;
    static int size = 0;
    static int[] dx = {0,0,1,-1};
    static int[] dy = {1,-1,0,0};
    static int[][] arr;
    
    static class Point{
        int x;
        int y;
        Point (int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
    
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        
        int cnt = 0; // 그림 개수
        int size = 0;
        
        arr = new int[n][m];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (arr[i][j] == 1) {
                    size = Math.max(size, bfs(i,j));
                    cnt++;
                }
            }
        }
        
        System.out.println(cnt);
        System.out.println(size);
        
        
    }
    
    private static int bfs(int x, int y) {
        Deque<Point> deq = new ArrayDeque<>();
        deq.addLast(new Point(x,y));
        int s = 1;
        arr[x][y] = 0;
        while (!deq.isEmpty()) {
            Point now = deq.pollFirst();
            
            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                if (nx < 0 || nx > (n-1) || ny < 0 || ny > (m-1)) continue;
                
                if (arr[nx][ny] == 1) {
                    arr[nx][ny] = 0;
                    s++;
                    deq.addLast(new Point(nx, ny));
                }
                
            }
        }
        
        return s;
    }
}